neicnordic/sensitive-data-archive
# htsget download endpoints

**Related issue(s) and PR(s)**  
This PR closes NBISweden/LocalEGA-SE-Deployment#725 and neicnordic/sda-download#372.

It is based on the work in https://github.com/neicnordic/sda-download/pull/327.

**Description**
- Adds endpoint `/s3-encrypted` which is similar to `/s3` but returns an encrypted file.
- Adds support for returning parts of encrypted files, either with url param `endCoordinate` or - for encrypted files - with the http header `Range: bytes=0-1000`. The later overrides the first, and is strictly interpreted, while `endCoordinate` will be used to find the nearest crypt4gh data block boundary and cut the file there.
- Minor modifications to work with htsget-rs (see https://github.com/GenomicDataInfrastructure/starter-kit-htsget/tree/feature/rust-htsget)

**How to test**
- get a token: `token=$(curl -s -k https://localhost:8080/tokens | jq -r '.[0]')`
- to download the complete file:  `curl 'http://localhost:8443/s3-encrypted/DATASET0001/htsnexus_test_NA12878.bam' -H "Authorization: Bearer $token" -o test.bam.c4gh `
- to download the first 3000 bytes:  `curl 'http://localhost:8443/s3-encrypted/DATASET0001/htsnexus_test_NA12878.bam' -H "Authorization: Bearer $token" -o test3000.bam.c4gh -H "Range: bytes=0-3000`
- to download the first 3000 bytes, and adding extra bytes to align to the data package boundaries :  `curl 'http://localhost:8443/s3-encrypted/DATASET0001/htsnexus_test_NA12878.bam?endCoordinate=3000' -H "Authorization: Bearer $token" -o test3000ok.bam.c4gh`
- to get the file info `curl --head -i 'http://localhost:8443/s3-encrypted/DATASET0001/htsnexus_test_NA12878.bam' -H "Authorization: Bearer $token"`

*Further comments*
- `startCoordinate` is not implemented (see neicnordic/sensitive-data-archive#696)
- also see neicnordic/sensitive-data-archive#400 and neicnordic/sensitive-data-archive#698 for future work that will be required
